zookeeper-server/src/main/java/org/apache/zookeeper/server/ZooKeeperServer.java

```diff
@@ -933,12 +933,23 @@ public void shutdown() {
      * @param fullyShutDown true if another server using the same database will not replace this one in the same process
      */
     public synchronized void shutdown(boolean fullyShutDown) {
+        if (shutdownZKServer(fullyShutDown)) {
+            updateZKDatabase(fullyShutDown);
+        }
+    }
+
+    /**
+     * Shut down the server instance
+     * @param fullyShutDown true if another server using the same database will not replace this one in the same process
+     * @return true if the server is successfully shutdown, false if the server cannot be shutdown.
+     */
+    public synchronized boolean shutdownZKServer(boolean fullyShutDown) {
```

Comment:
Will these two new methods be `protected`? I don't think they are intended to be called by app users.